### PR TITLE
egctl: implement `config envoy-proxy all` command

### PR DESCRIPTION
**Pull request by zirain**
xref: #915

### CODE REVIEW

**Comment by arkodg** [comment]
shouldnt the title be `egctl config envoy-proxy all <instance_name>`

File: internal/cmd/egctl/config.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	cfgCommand.AddCommand(proxyCommand())
+
+	flags := cfgCommand.Flags()
+	options.AddKubeConfigFlags(flags)
+
+	cfgCommand.PersistentFlags().StringVarP(&output, "output", "o", "json", "One of 'yaml' or 'json'")
+	cfgCommand.PersistentFlags().StringVar(&containerName, "container-name", "envoy", "Name of container where envoy proxy is installed.")
+	cfgCommand.PersistentFlags().Int32Var(&adminPort, "admin-port", 19000, "Admin port of envoy proxy.")
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:45]: lets rm this until we make it configurable ?

File: internal/cmd/egctl/config.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		Long:    "Retrieve information about proxy configuration from envoy proxy and gateway.",
+	}
+
+	cfgCommand.AddCommand(proxyCommand())
+
+	flags := cfgCommand.Flags()
+	options.AddKubeConfigFlags(flags)
+
+	cfgCommand.PersistentFlags().StringVarP(&output, "output", "o", "json", "One of 'yaml' or 'json'")
+	cfgCommand.PersistentFlags().StringVar(&containerName, "container-name", "envoy", "Name of container where envoy proxy is installed.")
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:44]: the original design doc had `instance-name`, why is `container-name` needed, can we do without adding this arg ?

File: internal/cmd/egctl/config.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+func allConfigCmd() *cobra.Command {
+
+	allConfigCmd := &cobra.Command{
+		Use:   "all <pod-name>",
+		Short: "Retrieves all configuration for the Envoy in the specified pod",
+		Long:  `Retrieve information about all configuration for the Envoy instance in the specified pod.`,
+		Example: `  # Retrieve summary about all configuration for a given pod from Envoy.
+  egctl config envoy-proxy all <pod-name> -n <pod-namespace>
+
+  # Retrieve full cluster dump as YAML
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:72]: why `cluster` ?

File: internal/cmd/egctl/config.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  egctl config envoy-proxy all <pod-name> -n <pod-namespace> -o yaml
+
+  # Retrieve full cluster dump with short syntax
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:75]: same as above

File: internal/cmd/egctl/config.go (hunk 1 above)
Comment (zirain) [reply internal/cmd/egctl/config.go:45]: this's my fault, I test this on an istio pod locally.

File: internal/cmd/egctl/config.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	c.AddCommand(allConfigCmd())
+
+	return c
+}
+
+func allConfigCmd() *cobra.Command {
+
+	allConfigCmd := &cobra.Command{
+		Use:   "all <pod-name>",
+		Short: "Retrieves all configuration for the Envoy in the specified pod",
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:68]:
```suggestion
		Short: "Retrieves all Envoy xDS resources from the specified pod",
```

File: internal/cmd/egctl/config.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		Long:  `Retrieve information about all configuration for the Envoy instance in the specified pod.`,
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:69]:
```suggestion
		Long:  `Retrieves information about all Envoy xDS resources from the specified pod.`,
```

File: internal/cmd/egctl/config.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	c.AddCommand(allConfigCmd())
+
+	return c
+}
+
+func allConfigCmd() *cobra.Command {
+
+	allConfigCmd := &cobra.Command{
```
Comment (arkodg) [inline internal/cmd/egctl/config.go:66]: thoughts on adding `-A` to get information from all pods in all namespaces ?

Comment (zirain) [reply internal/cmd/egctl/config.go:66]: I never thought of such a user case, can you share more details?
it would be better to track that with a ticket.

Comment (Xunzhuo) [reply internal/cmd/egctl/config.go:66]: I think it makes sense, `-A` allows users to get info from all pods in all namespaces, like `kubectl get <resources> -A`.

File: internal/cmd/egctl/config.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		Short: "Retrieves all Envoy xDS resources from the specified pod",
+		Long:  `Retrieves information about all Envoy xDS resources from the Envoy instance in the specified pod.`,
+		Example: `  # Retrieve summary about all configuration for a given pod from Envoy.
+  egctl config envoy-proxy all <pod-name> -n <pod-namespace>
+
+  # Retrieve full configuration dump as YAML
+  egctl config envoy-proxy all <pod-name> -n <pod-namespace> -o yaml
+
+  # Retrieve full configuration dump with short syntax
+  egctl c proxy a <pod-name> -n <pod-namespace>
```
Comment (Xunzhuo) [inline internal/cmd/egctl/config.go:77]: I feel this is a bit weird here, The short name `a` here means `all`, but in an english word, it represents a single odd number.

File: internal/cmd/egctl/config.go (hunk 7 above)
Comment (arkodg) [reply internal/cmd/egctl/config.go:66]: yah helps users like me who dont want to run an extra `kubectl get po -A` cmd :)  to get pod name

File: internal/cmd/egctl/config.go (hunk 8 above)
Comment (arkodg) [reply internal/cmd/egctl/config.go:77]: agreed, we dont need a short name for `all`

Comment (zirain) [reply internal/cmd/egctl/config.go]: reverted

**Comment by arkodg** [comment]
thanks @zirain this looks like a great start, hoping you can add some test cases since the current diff coverage for this PR is 0%

File: internal/cmd/egctl/config_test.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+		},
+		{
+			output:   "yaml",
+			expected: "out.yaml",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.output, func(t *testing.T) {
+			got, err := extractConfigDump(fw, tc.output)
```
Comment (Xunzhuo) [inline internal/cmd/egctl/config_test.go:95]: Thanks for your effort to put UT around this feature. But in my view, this kind of testes could be better to add as e2e tests. 

In this approach, we put configDumps into fake server, and request to it and get configDumps back, and compare them. It make less sence I think 😄. And we will contain long test jsons/yamls to mimic real envoyproxy configs. cc @arkodg WDYT.

Comment (zirain) [reply internal/cmd/egctl/config_test.go:95]: it's useful for `config envoy-proxy cluster`

Comment (arkodg) [reply internal/cmd/egctl/config_test.go:95]: @Xunzhuo yeah e2e would be ideal, maybe we can raise a GH issue to track it, but for now, this approach add coverage for a sufficient amount of code in this PR, which Im fine with

File: tools/linter/codespell/.codespell.skip [hunk 10]
```diff
@@ -12,3 +12,4 @@
 go.mod
 go.sum
 bin
+testdata
```
Comment (arkodg) [inline tools/linter/codespell/.codespell.skip:15]: this isn't right, we were linting `testdata` earlier

File: internal/cmd/egctl/config_test.go (hunk 9 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config_test.go:95]: Thanks @zirain, @arkodg. It makes sense.

File: tools/linter/codespell/.codespell.skip (hunk 10 above)
Comment (zirain) [reply tools/linter/codespell/.codespell.skip:15]: let me ignore special file, configdump file will not pass codespell check.

Comment (zirain) [reply tools/linter/codespell/.codespell.skip]: done

**Review by Xunzhuo — DISMISSED** [review]
Thanks @zirain

File: tools/linter/codespell/.codespell.skip (hunk 10 above)
Comment (arkodg) [reply tools/linter/codespell/.codespell.skip]: I pulled in your branch and deleted the `tracer` that was causing the issue
```
diff --git a/internal/cmd/egctl/testdata/config/out.json b/internal/cmd/egctl/testdata/config/out.json
index d7bc3d3..1bd6f16 100644
--- a/internal/cmd/egctl/testdata/config/out.json
+++ b/internal/cmd/egctl/testdata/config/out.json
@@ -1316,10 +1316,6 @@
                                 "envoy.extensions.filters.network.thrift_proxy.router.v3.Router"
                             ]
                         },
-                        {
-                            "name": "envoy.dynamic.ot",
-                            "category": "envoy.tracers"
-                        },

```
this fixed `codespell`

**Review by arkodg — APPROVED** [review]
LGTM, thanks !

**Pull request merged**